core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/MSYS2.java

```diff
@@ -0,0 +1,164 @@
+/********************************************************************************
+ * Copyright (c) 2022 徐持恒 Xu Chiheng
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ********************************************************************************/
+package org.eclipse.cdt.utils;
+
+import java.io.File;
+import java.util.Map;
+
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.core.runtime.Platform;
+
+// A collection of MSYS2-related utilities.
+public class MSYS2 {
+	public static boolean present;
+	private static String rootDir;
+	static {
+		initialize();
+	}
+
+	// initialize static data fields
+	private static void initialize() {
+		if (Platform.getOS().equals(Platform.OS_WIN32)) {
+			Map<String, String> environment = System.getenv();
+			rootDir = environment.get("MSYS2_DIR"); //$NON-NLS-1$
+			if (rootDir != null) {
+				if (dirHasMsys2Dll(rootDir)) {
+					present = true;
+					rootDir = new Path(rootDir).toPortableString();
+					return;
+				}
+			}
+			for (char drive = 'C'; drive < 'H'; drive++) {
+				StringBuilder dirStringBuilder = new StringBuilder();
+				dirStringBuilder.append(drive);
+				dirStringBuilder.append(":/msys64"); //$NON-NLS-1$
+				String dirString = dirStringBuilder.toString();
+				if (dirHasMsys2Dll(dirString)) {
+					present = true;
+					rootDir = new Path(dirString).toPortableString();
+					return;
+				}
+			}
+		}
+	}
+
+	private static boolean dirHasMsys2Dll(String dirString) {
+		IPath dirLocation = new Path(dirString);
+		File dir = dirLocation.toFile();
+		if (dir.isAbsolute() && dir.exists() && dir.isDirectory()) {
+			File file = dirLocation.append("/usr/bin/msys-2.0.dll").toFile(); //$NON-NLS-1$
+			if (file.exists() && file.isFile() && file.canRead()) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	public boolean isPresent() {
+		return present;
+	}
+
+	public String getRootDir() {
+		return rootDir;
+	}
+
+	// Convert Unix path to Windows path
+	public static String pathToWindows(String unixPath) {
+		if (!present) {
+			return unixPath;
+		}
+		if (unixPath == null || unixPath.trim().length() == 0) {
+			return unixPath;
+		}
+
+		String windowsPath;
+		IPath path = Path.fromOSString(unixPath);
+		if (path.getDevice() != null) {
+			// already a windows absolute path
```

Comment:
A device doesn't mean an absolute path on Windows. There is Path.isAbsolute() that will confirm that. Also, not sure if/how this code handles UNC, but perhaps that is out of scope for now.

e.g. `c:file.c` will be identified as absolute according to this code, but it refers to the `file.c` in the cwd of the C: drive.